charts/generic-service/templates/controller.yaml

```
@@ -51,59 +58,56 @@ spec:
     {{- end }}
     type: OnDelete
   {{- else if eq .Values.rollout.strategy "Canary" }}
-    {{- if not .Values.ingress.enabled }}
-      {{ fail "ingress.enabled must be true if rollout.strategy is Canary" }}
-    {{- end }}
-    {{- if not .Values.ingress.domains }}
-      {{ fail "ingress.domains must not be empty if rollout.strategy is Canary" }}
-    {{- end }}
     {{- if eq .Values.rollout.controller "ArgoRollout" }}
     canary:
-      stableService: {{ include "generic-service.fullname" . }}
-      stableMetadata:
-        labels:
-          role: stable
-      canaryService: {{ include "generic-service.fullname" . }}-preview
-      canaryMetadata:
-        labels:
-          role: preview
-      analysis: {{- .Values.rollout.analysis | required "rollout.analysis is required if rollout.strategy is Canary" | toYaml |  nindent 8 }}
-      trafficRouting:
-        {{- if .Values.ingress.istio.enabled }}
-        istio:
-          virtualService:
-            name: {{ include "generic-service.fullname" . }}
-        {{- else if contains "nginx" .Values.ingress.class }}
-        nginx:
-          stableIngress: {{ include "generic-service.fullname" . }}
-        {{- else }}
-          {{ fail "Ingress must use Istio or nginx if rollout.strategy is Canary" }}
+      {{- with .Values.rollout.analysis -}}
+        analysis: {{- . | toYaml | nindent 8 }}
+      {{- end }}
```

`with` doesn't seem necessary here.

Indent seems to be too deep. Would put `analysis` under `trafficRouting`.

```suggestion
      analysis: {{- .Values.rollout.analysis | toYaml | nindent 8 }}
```

---

`analysis` is an optional field (with or without traffic routing). see [rollout specs](https://argo-rollouts.readthedocs.io/en/stable/features/specification/)

---

Then perhaps like this?

```suggestion
      {{- if.Values.rollout.analysis }}
      analysis: {{- . | toYaml | nindent 8 }}
      {{- end }}
```